File: 107_PopulatingNextRightPointers.cpp
// Topic: Populating Next Right Pointers in Each Node
// Description: Given a perfect binary tree, connect each node's next pointer to its next right
// node. Approach: Use level traversal or next pointers to iterate O(1) space (excluding recursion
// stack). We'll use iterative level-by-level without extra containers.

#include <bits/stdc++.h>
using namespace std;

struct Node
{
    int val;
    Node* left;
    Node* right;
    Node* next;
    Node(int x) : val(x), left(nullptr), right(nullptr), next(nullptr) {}
};

Node* buildPerfectTreeLevel(const vector<int>& vals)
{
    if (vals.empty())
        return nullptr;
    int n = vals.size();
    vector<Node*> nodes(n);
    for (int i = 0; i < n; ++i) nodes[i] = (vals[i] == -1 ? nullptr : new Node(vals[i]));
    int idx = 0;
    for (int i = 0; i < n && nodes[i]; ++i)
    {
        int left = 2 * i + 1, right = 2 * i + 2;
        if (left < n)
            nodes[i]->left = nodes[left];
        if (right < n)
            nodes[i]->right = nodes[right];
    }
    return nodes[0];
}

Node* connect(Node* root)
{
    if (!root)
        return root;
    Node* leftmost = root;
    while (leftmost->left)
    {
        Node* head = leftmost;
        while (head)
        {
            head->left->next = head->right;
            if (head->next)
                head->right->next = head->next->left;
            head = head->next;
        }
        leftmost = leftmost->left;
    }
    return root;
}

void printLevels(Node* root)
{
    Node* level = root;
    while (level)
    {
        Node* cur = level;
        while (cur)
        {
            cout << cur->val;
            if (cur->next)
                cout << ' ';
            cur = cur->next;
        }
        cout << "\n";
        level = level->left;
    }
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    vector<vector<int>> tests = {{1, 2, 3, 4, 5, 6, 7}, {1}, {}};

    for (int i = 0; i < (int)tests.size(); ++i)
    {
        Node* root = buildPerfectTreeLevel(tests[i]);
        connect(root);
        cout << "Test " << (i + 1) << ":\n";
        printLevels(root);
    }
    return 0;
}

/*
Time Complexity: O(n)
Space Complexity: O(1) extra (excluding tree)
*/

File: 107_PopulatingNextRightPointers.py
# Topic: Populating Next Right Pointers in Each Node
# Description: Given a perfect binary tree, connect each node's next pointer to its next right node.
# Approach: Use level traversal or next pointers to iterate O(1) space. Iterative level-by-level.

from typing import Optional, List


class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next


def build_perfect_tree(vals: List[int]) -> Optional[Node]:
    if not vals:
        return None
    nodes = [Node(v) if v != -1 else None for v in vals]
    for i, node in enumerate(nodes):
        if node:
            left = 2 * i + 1
            right = 2 * i + 2
            if left < len(nodes):
                node.left = nodes[left]
            if right < len(nodes):
                node.right = nodes[right]
    return nodes[0] if nodes else None


class Solution:
    def connect(self, root: Optional[Node]) -> Optional[Node]:
        if not root:
            return root
        leftmost = root
        while leftmost.left:
            head = leftmost
            while head:
                head.left.next = head.right
                if head.next:
                    head.right.next = head.next.left
                head = head.next
            leftmost = leftmost.left
        return root


def print_levels(root: Optional[Node]) -> None:
    level = root
    while level:
        cur = level
        vals = []
        while cur:
            vals.append(str(cur.val))
            cur = cur.next
        print(' '.join(vals))
        level = level.left


if __name__ == "__main__":
    tests = [[1, 2, 3, 4, 5, 6, 7], [1], []]
    for i, vals in enumerate(tests):
        root = build_perfect_tree(vals) if vals else None
        Solution().connect(root)
        print(f"Test {i + 1}:")
        if root:
            print_levels(root)
        else:
            print("Empty")

"""
Time Complexity: O(n)
Space Complexity: O(1) extra (excluding tree)
"""
